// @pcmnac/single-spa-cycle/lib/single-spa-cycle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.default = singleSpaCycle;
var defaultOpts = {
	// required opts
	run: null,
	drivers: null,
	rootComponent: null
};

function singleSpaCycle(userOpts) {
	if ((typeof userOpts === 'undefined' ? 'undefined' : _typeof(userOpts)) !== 'object') {
		throw new Error('single-spa-cycle requires a configuration object');
	}

	var opts = _extends({}, defaultOpts, userOpts);

	if (!opts.run) {
		throw new Error('single-spa-cycle must be passed opts.run function');
	}

	if (!opts.drivers) {
		throw new Error('single-spa-cycle must be passed opts.drivers');
	}

	if (!opts.rootComponent) {
		throw new Error('single-spa-cycle must be passed opts.rootComponent');
	}

	return {
		bootstrap: bootstrap.bind(null, opts),
		mount: mount.bind(null, opts),
		unmount: unmount.bind(null, opts)
	};
}

function bootstrap(opts) {
	return Promise.resolve();
}

function mount(opts) {
	return new Promise(function (resolve, reject) {
		opts.dispose = opts.run(opts.rootComponent, opts.drivers);
		resolve();
	});
}

function unmount(opts) {
	return new Promise(function (resolve, reject) {
		if (opts.dispose) {
			opts.dispose();
			opts.dispose = null;
		}
		resolve();
	});
}
